node/metered-channel/src/oneshot.rs

```diff
@@ -0,0 +1,348 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Metered variant of oneshot channels to be able to extract delays caused by delayed responses.
+
+use std::{
+	convert::TryFrom,
+	pin::Pin,
+	sync::{
+		atomic::{AtomicU64, AtomicU8, Ordering},
+		Arc,
+	},
+	task::{Context, Poll},
+	time::{Duration, Instant},
+};
+
+use futures::{
+	channel::oneshot::{self, Canceled, Cancellation},
+	future::{Fuse, FusedFuture},
+	prelude::*,
+};
+use futures_timer::Delay;
+use num_enum::{IntoPrimitive, TryFromPrimitive};
+
+/// Provides the reason for termination, or [`Self::Unfinished`] if none exists.
+#[derive(Debug, Clone, Copy, IntoPrimitive, TryFromPrimitive, PartialEq, Eq)]
+#[repr(u8)]
+pub enum Reason {
+	Unfinished = 0,
+	Compeletion = 1,
+	Cancellation = 2,
+	HardTimeout = 3,
+}
+
+/// A peek into delay and state of a `oneshot` channel.
+#[derive(Debug, Clone, Default)]
+pub struct Meter {
+	/// Delay between first poll and completion or cancellation.
+	first_poll_till_end_in_millis: Arc<AtomicU64>,
+	/// Determine the reason of completion.
+	reason: Arc<AtomicU8>,
```

Comment:
Doesn't seem like we need two reference counters.
I would suggest extracting the fields into a struct and then meter would contain one `Arc<MeterInner>` instead of two.